flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/ChangelogStreamHandleReaderWithCacheTest.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.changelog.fs;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.CoreOptions;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.state.filesystem.FileStateHandle;
+import org.apache.flink.runtime.state.memory.ByteStreamStateHandle;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import java.io.DataInputStream;
+import java.time.Duration;
+import java.util.UUID;
+
+import static org.apache.flink.changelog.fs.FsStateChangelogOptions.CACHE_IDLE_TIMEOUT;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/** {@link ChangelogStreamHandleReaderWithCache} test. */
+class ChangelogStreamHandleReaderWithCacheTest {
+
+    @TempDir java.nio.file.Path tempFolder;
+
+    @Test
+    void testCloseStreamTwice() throws Exception {
+        String path = tempFolder.toUri().getPath();
+
+        byte[] bytes = {0x00}; // not compressed, empty data
+        ByteStreamStateHandle byteStreamHandle =
+                new ByteStreamStateHandle(UUID.randomUUID().toString(), bytes);
+
+        FileStateHandle stateHandler = mock(FileStateHandle.class);
+        when(stateHandler.openInputStream()).thenReturn(byteStreamHandle.openInputStream());
+
+        FileSystem mockFileSystem = mock(FileSystem.class);
+        when(mockFileSystem.isDistributedFS()).thenReturn(true);
```

Comment:
Thanks for adding the test!

Have you considered creating real objects instead of mocks?

In the past, tests with mocks were quite fragile.
And there are a lot of calls in this test, so it can become fragile too.

WDYT?

Reply:
Thanks @rkhachatryan , I updated the pr and use real objects instead of mocks. But I'm confused about "tests with mocks are fragile", can you explain further ?

BTW, I squash all commits to one.

Reply:
Thanks a lot @zoltar9264!

For mocked objects, expected sequence of calls is usually specified. However, sequence of calls is not a public contract and may change in the future. That might result in NPE for example, if a new invocation is added and the result is used.

Reply:
Thanks for the clarification @rkhachatryan !